core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/GroupByAggregator.java

```diff
@@ -0,0 +1,182 @@
+    for (String groupByField : groupByFields) {
+      Schema.Field field = inputSchema.getField(groupByField);
+      if (field == null) {
+        throw new IllegalArgumentException(String.format(
+          "Cannot group by field '%s' because it does not exist in input schema %s.",
+          groupByField, inputSchema));
+      }
+      outputFields.add(field);
+    }
+
+    // check that all fields needed by aggregate functions exist in the input schema.
+    for (GroupByConfig.FunctionInfo functionInfo : aggregates) {
+      // special case count(*) because we don't have to check that the input field exists
+      if (functionInfo.getField().equals("*")) {
+        AggregateFunction aggregateFunction = functionInfo.getAggregateFunction(null);
+        outputFields.add(Schema.Field.of(functionInfo.getName(), aggregateFunction.getOutputSchema()));
+        continue;
+      }
+
+      Schema.Field inputField = inputSchema.getField(functionInfo.getField());
+      if (inputField == null) {
+        throw new IllegalArgumentException(String.format(
+          "Invalid aggregate %s(%s): Field '%s' does not exist in input schema %s.",
+          functionInfo.getFunction(), functionInfo.getField(), functionInfo.getField(), inputSchema));
+      }
+      AggregateFunction aggregateFunction = functionInfo.getAggregateFunction(inputField.getSchema());
+      outputFields.add(Schema.Field.of(functionInfo.getName(), aggregateFunction.getOutputSchema()));
+    }
+    return Schema.recordOf(inputSchema.getRecordName() + ".agg", outputFields);
+  }
+
+  private void updateAggregates(StructuredRecord groupVal) {
+    for (AggregateFunction aggregateFunction : aggregateFunctions.values()) {
+      aggregateFunction.update(groupVal);
+    }
+  }
+
+  private void initAggregates(Schema valueSchema) {
+    List<Schema.Field> outputFields = new ArrayList<>(groupByFields.size() + functionInfos.size());
+    for (String groupByField : groupByFields) {
+      outputFields.add(valueSchema.getField(groupByField));
+    }
+
+    aggregateFunctions = new HashMap<>();
+    for (GroupByConfig.FunctionInfo functionInfo : functionInfos) {
+      Schema.Field inputField = valueSchema.getField(functionInfo.getField());
+      Schema fieldSchema = inputField == null ? null : inputField.getSchema();
+      AggregateFunction aggregateFunction = functionInfo.getAggregateFunction(fieldSchema);
+      aggregateFunction.beginAggregate();
+      outputFields.add(Schema.Field.of(functionInfo.getName(), aggregateFunction.getOutputSchema()));
+      aggregateFunctions.put(functionInfo.getName(), aggregateFunction);
+    }
+    outputSchema = Schema.recordOf(valueSchema.getRecordName() + ".agg", outputFields);
+  }
+
+  private Schema getGroupKeySchema(Schema inputSchema) {
+    List<Schema.Field> fields = new ArrayList<>();
+    for (String groupByField : conf.getGroupByFields()) {
+      fields.add(inputSchema.getField(groupByField));
```

Comment:
`inputSchema.getField(groupByField)` could be null right? is it okay to add that to the list?